File: cplusplus/CoroServe/boring_client.cc
#include "net/net.hh"
#include "examples.hh"

namespace boring_client {

using coro::net::Socket;

void launch() {
  coro::net::Initializer<> init;

  std::unique_ptr <Socket, std::function<void (Socket*)>> c{
    new Socket{}, [](Socket* s) { s->close(); }
  };

  if (!c->start_connect())
    return;

  for (;;) {
    char rbuf[1024]{};
    if (auto n = c->read(rbuf, sizeof(rbuf)); n > 0)
      std::cout << "BORING CLIENT: recv: " << rbuf << std::endl;
    else
      break;
  }
}

}

CORO_EXAMPLE(BoringClient, bc, "a boring client") {
  boring_client::launch();
}

File: cplusplus/CoroServe/boring_client.go
// Copyright (c) 2020 ASMlover. All rights reserved.
//
// Redistribution and use in source and binary forms, with or without
// modification, are permitted provided that the following conditions
// are met:
//
//  * Redistributions of source code must retain the above copyright
//    notice, this list ofconditions and the following disclaimer.
//
//  * Redistributions in binary form must reproduce the above copyright
//    notice, this list of conditions and the following disclaimer in
//    the documentation and/or other materialsprovided with the
//    distribution.
//
// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
// "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
// LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
// FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
// COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
// INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
// BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
// LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
// CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
// LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
// ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
// POSSIBILITY OF SUCH DAMAGE.

package main

import (
	"fmt"
	"net"
)

func boringClient(conn net.Conn) {
	for {
		var buf [1024]byte
		n, err := conn.Read(buf[0:])
		if err != nil {
			break
		}
		fmt.Println("BORING CLIENT: recv: ", string(buf[0:n]))
	}
	conn.Close()
}

func main() {
	conn, err := net.Dial("tcp", "127.0.0.1:5555")
	if err != nil {
		return
	}
	boringClient(conn)
}
